sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -184,11 +184,22 @@ private[sql] object Dataset {
  */
 @Stable
 class Dataset[T] private[sql](
-    @transient val sparkSession: SparkSession,
+    @transient private val _sparkSession: SparkSession,
     @DeveloperApi @Unstable @transient val queryExecution: QueryExecution,
     @DeveloperApi @Unstable @transient val encoder: Encoder[T])
   extends Serializable {
 
+  @transient lazy val sparkSession: SparkSession = {
+    if (_sparkSession == null) {
+      throw new SparkException(
+      "Dataset transformations and actions can only be invoked by the driver, not inside of" +
+        " other transformations; for example, dataset1.map(x => dataset2.values.count() * x)" +
```

Comment:
While I agree with this in general, `not inside of other transformations` can be controversial: e.g. `Dataset.transform`. I would just remove this words or reword it.

Reply:
What about `not inside of other Dataset transformations`?

Reply:
@HyukjinKwon 
I updated message